src/flowMC/sampler/HMC.py

```diff
@@ -0,0 +1,173 @@
+            log_prob = jnp.where(do_accept, proposed_ham, H)
+
+            return position, log_prob, do_accept
+        
+        if return_aux == False:
+            return hmc_kernel
+        else:
+            return hmc_kernel, leapfrog_kernal, leapfrog_step
+            
+
+    def make_update(self) -> Callable:
+        """
+        
+        Making HMC update function for multiple steps
+
+        """
+
+        hmc_kernel = self.make_kernel()
+
+        def hmc_update(i, state):
+            key, positions, log_Ham, acceptance = state
+            _, key = jax.random.split(key)
+            new_position, new_log_Ham, do_accept = hmc_kernel(
+                key, positions[i - 1], log_Ham[i - 1]
+            )
+            positions = positions.at[i].set(new_position)
+            log_Ham = log_Ham.at[i].set(new_log_Ham)
+            acceptance = acceptance.at[i].set(do_accept)
+            return (key, positions, log_Ham, acceptance)
+
+        return hmc_update
+
+    def make_sampler(self) -> Callable:
+        hmc_update = self.make_update()
+
+        if self.jit:
+            hmc_update = jax.jit(hmc_update)
+
+        hmc_update = jax.vmap(hmc_update, in_axes = (None, (0, 0, 0, 0)), out_axes=(0, 0, 0, 0))
+        lh = jax.vmap(self.get_initial_hamiltonian)
+
+        def hmc_sampler(rng_key, n_steps, initial_position):
+            
+            keys = jax.vmap(jax.random.split)(rng_key)
+            rng_key = keys[:, 0]
+            rng_init = keys[:, 1]
+            logp = lh(rng_init,initial_position)
+            n_chains = rng_key.shape[0]
+            acceptance = jnp.zeros(
+                (
+                    n_chains,
+                    n_steps
+                )
+            )
+            all_positions = (
+                jnp.zeros(
+                    (
+                        n_chains,
+                        n_steps,
+                    )
+                    + initial_position.shape[-1:]
+                )
+                + initial_position[:, None]
+            )
+            all_logp = (
+                jnp.zeros(
+                    (
+                        n_chains,
+                        n_steps,
+                    )
+                )
+                + logp[:, None]
+            )
+            state = (rng_key, all_positions, all_logp, acceptance)
+            for i in tqdm(
+                range(1, n_steps), desc="Sampling Locally", miniters=int(n_steps / 10)
+            ):
+                state = hmc_update(i, state)
+            return state
+
+        return hmc_sampler
+
+    def make_leapforg_kernel(self):
+        def leapfrog_kernal(carry, data):
+            position, momentum = carry
+            position = position + self.step_size * self.grad_kinetic(momentum)
+            momentum = momentum - self.step_size * self.grad_potential(position)
+            return position, momentum
+        return leapfrog_kernal
+
+
+
+
+
+from tqdm import tqdm
```

Comment:
```suggestion

```

Should we get rid of the two last lines or are they around because that will be used in the autotuning to be implemented later?

Reply:
They are there for autotune implementation later.